modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/rel/AsyncRootNode.java

```diff
@@ -255,11 +253,17 @@ private void flush() throws Exception {
             currentReq.fut.complete(new BatchedResult<>(currentReq.buff, hasMore == HasMore.YES));
         }
 
-        if (waiting == 0 && buff.isEmpty()) {
-            //noinspection NestedAssignment
-            source.request(waiting = IN_BUFFER_SIZE);
-        } else if (hasMore == HasMore.NO) {
-            closeAsync();
+        if (buff.isEmpty()) {
+            if (waiting == 0) {
+                //noinspection NestedAssignment
+                source.request(waiting = IN_BUFFER_SIZE);
+            } else if (waiting == -1) {
+                assert hasMore == HasMore.NO : hasMore;
+
+                closeAsync();
+            }
+        } else if (!pendingRequests.isEmpty()) {
+            source.context().execute(this::flush, source::onError);
         }
```

Comment:
Why do you enqueue more than one flush task?
Should it be
```suggestion
        } else if (!pendingRequests.isEmpty()) {
            scheduleTask();
        }
```

Reply:
nice catch! fixed